RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -116,156 +140,367 @@ void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const ed
     
-  EcalClusterLazyTools lazyTool(evt, es, _recHitsEB, _recHitsEE);
-  
-  //magic numbers for MINUIT-like transformation of BDT output onto limited range
-  //(These should be stored inside the conditions object in the future as well)
-  const double meanlimlow = -0.336;
-  const double meanlimhigh = 0.916;
-  const double meanoffset = meanlimlow + 0.5*(meanlimhigh-meanlimlow);
-  const double meanscale = 0.5*(meanlimhigh-meanlimlow);
-  
-  const double sigmalimlow = 0.001;
-  const double sigmalimhigh = 0.4;
-  const double sigmaoffset = sigmalimlow + 0.5*(sigmalimhigh-sigmalimlow);
-  const double sigmascale = 0.5*(sigmalimhigh-sigmalimlow);  
-  
   for (unsigned int idx = 0; idx<cs.size(); ++idx) {
     
     reco::PFCluster &cluster = cs[idx];
     
     double e = cluster.energy();
-    double pt = cluster.pt(); 
-    
+    double pt = cluster.pt();
+   
     //limit raw energy value used to evaluate corrections
     //to avoid bad extrapolation
     double evale = e;
-    if (_maxPtForMVAEvaluation>0. && pt>_maxPtForMVAEvaluation) {
-      evale *= _maxPtForMVAEvaluation/pt; 
+    if (maxPtForMVAEvaluation_ > 0. && pt > maxPtForMVAEvaluation_) {
+      evale *= maxPtForMVAEvaluation_/pt; 
     }
     
     double invE = (e == 0.) ? 0. : 1./e; //guard against dividing by 0.
     
-    int size = lazyTool.n5x5(cluster);
+    bool iseb = (cluster.layer() == PFLayer::ECAL_BARREL);
+    int clusFlag = 0;
     
-    bool iseb = cluster.layer() == PFLayer::ECAL_BARREL;
+    if (iseb){
+      EBSrFlagCollection::const_iterator srf = ebSrFlags->find(readOutUnitOf(static_cast<EBDetId>(cluster.seed())));      
 
-    //find index of corrections (0-4 for EB, 5-9 for EE, depending on cluster size and raw pt)
-    int coridx = std::min(size,3)-1;
-    if (coridx==2) {
-      if (pt>4.5) {
-        coridx += 1;
-      }
-      if (pt>18.) {
-        coridx += 1;
-      }
+      if (srf != ebSrFlags->end())
+	clusFlag = srf->value();
+      else
+	clusFlag = 3; ///if the flag of that seed xtal is not found - just assume full readout
+      
+    } else {
+      EESrFlagCollection::const_iterator srf = eeSrFlags->find(readOutUnitOf(static_cast<EEDetId>(cluster.seed())));      
+
+      if (srf != eeSrFlags->end())
+	clusFlag = srf->value();
+      else
+	clusFlag = 3;
+      
     }
-    if (!iseb) {
-      coridx += 5;
+  
+
+
+    //find index of corrections (0-3 for EB, 4-7 for EE, depending on cluster size and raw pt)
+    int coridx = 0;
+    int regind = 0;
+
+    if (!iseb) regind = 4;
+
+    if (clusFlag==1)
+      coridx = 0 + regind;
+    else {
+      if (pt<2.5) coridx = 1 + regind;
+      else if (pt>=2.5 && pt<6.) coridx = 2 + regind;
+      else if (pt>=6.) coridx = 3 + regind;
     }
-    
+    if (clusFlag!=1 && clusFlag!=3) {
+      edm::LogWarning("PFClusterEMEnergyCorrector") << "We can only correct regions readout in ZS (flag 1) or FULL readout (flag 3). Flag " << clusFlag << " is not recognized."
```

Comment:
It is not supposed to happen at all, unless ECAL changes the way they apply ZS/SR. I will put as LogInfo. I think it is important for the user to know that something has changed, but I don't want to flood grid jobs.